### PR TITLE
feat(route53): Domain redirect pattern

### LINKED ISSUE #3893: Add a construct to redirect traffic to another domain with HTTPS support
### General Information

* [x] :wave: I **am able** to implement this feature request

* [ ] :warning: This feature might incur a breaking change

### Description

Add a construct to redirect internet traffic to another domain and redirect HTTP requests to HTTPS

### Proposed Solution

Implement the following https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/tutorial-redirecting-dns-queries.html using Route53, CloudFront, ACM and S3.

```ts
new HTTPSRedirect(this, 'Redirect', {
  fromDomain: 'example.com',
  toDomain: 'example.net',
  hostedZone: myHostedZone, // to create the record
});
```

To which package does this belong? I would say `@aws-cdk/aws-route53` or `@aws-cdk/aws-cloudfront`...

### PR DESCRIPTION
**Pull request by hoegertn**
This implements an HTTPS redirect using a CloudFront distribution and an S3 bucket. (fix #3893)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by hoegertn** [comment]
This is currently work-in-progress but I wanted to get feedback early in the process.

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  readonly zone: IHostedZone;
+  /**
+   * The redirect target
+   */
+  readonly redirectTarget: string;
+  /**
+   * The domain name
+   *
+   * @default - the domain name of the zone
+   */
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Should this only be a subdomain? Will this also support redirecting the root domain?

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import { Construct, Fn, RemovalPolicy } from '@aws-cdk/core';
+
+export interface HTTPSRedirectProps {
+  /**
+   * HostedZone of the domain
+   */
+  readonly zone: IHostedZone;
+  /**
+   * The redirect target
+   */
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Rename to targetDomain

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+export class HTTPSRedirect extends Construct {
+  constructor(scope: Construct, id: string, props: HTTPSRedirectProps) {
+    super(scope, id);
+
+    const domainName = props.domainName || props.zone.zoneName;
+
+    const redirectCertArn = props.certificate ? props.certificate.certificateArn : new DnsValidatedCertificate(this, 'RedirectCertificate', {
+      domainName,
+      hostedZone: props.zone,
+      region: 'us-east-1',
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Will this work if the stack is not in us-east-1? If not, we should error.

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        hostName: props.redirectTarget,
+        protocol: RedirectProtocol.HTTPS,
+      },
+      removalPolicy: RemovalPolicy.DESTROY,
+    });
+    const redirectDist = new CloudFrontWebDistribution(this, 'RedirectDistribution', {
+      originConfigs: [{
+        behaviors: [{ isDefaultBehavior: true }],
+        customOriginSource: {
+          domainName: Fn.select(2, Fn.split('/', redirectBucket.bucketWebsiteUrl)),
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Are you sure there isn’t an attribute for this already?

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+          originProtocolPolicy: OriginProtocolPolicy.HTTP_ONLY,
+        },
+      }],
+      aliasConfiguration: {
+        acmCertRef: redirectCertArn,
+        names: [domainName],
+      },
+      comment: `${domainName} Redirect to ${props.redirectTarget}`,
+      priceClass: PriceClass.PRICE_CLASS_ALL,
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Should this be configurable?

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Perhaps call it recordName to align with route53 terminology? Provide an @example

File: packages/@aws-cdk/aws-route53-patterns/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+> compatible changes or removal in any future version. Use of the API is not recommended in production
+> environments. Experimental APIs are not subject to the Semantic Versioning model.
+
+---
+<!--END STABILITY BANNER-->
+
+This library contains commonly used patterns for Route53:
+* HTTP Redirect
+  ```ts
+  new HTTPSRedirect(stack, 'Redirect', {
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/README.md]: Should be called `HttpsRedirect`

File: packages/@aws-cdk/aws-route53-patterns/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+> This API is still under active development and subject to non-backward
+> compatible changes or removal in any future version. Use of the API is not recommended in production
+> environments. Experimental APIs are not subject to the Semantic Versioning model.
+
+---
+<!--END STABILITY BANNER-->
+
+This library contains commonly used patterns for Route53:
+* HTTP Redirect
+  ```ts
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/README.md]: Provide some details about what this construct does and a glimpse to the underlying implementation in a sentence.

**Comment by eladb** [comment]
@rix0rrr what do you think about the module? Does it make sense to file this under "route53"? I am not 100% convinced.

File: packages/@aws-cdk/aws-route53-patterns/README.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+>
+> This API is still under active development and subject to non-backward
+> compatible changes or removal in any future version. Use of the API is not recommended in production
+> environments. Experimental APIs are not subject to the Semantic Versioning model.
+
+---
+<!--END STABILITY BANNER-->
+
+This library contains commonly used patterns for Route53:
+* HTTP Redirect
```
Comment (jogold) [inline packages/@aws-cdk/aws-route53-patterns/README.md]:
```suggestion
* HTTPS Redirect
```

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 3 above)
Comment (jogold) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: It works even if the stack is not deployed in `us-east-1`

https://github.com/aws/aws-cdk/blob/cbf41725bb13c5f75d1385cb70945fa8d7e3ce9c/packages/%40aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts#L18-L25

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  readonly redirectTarget: string;
+  /**
+   * The domain name
+   *
+   * @default - the domain name of the zone
+   */
+  readonly domainName?: string;
+  /**
+   * The ARN of the certificate; Has to be in us-east-1
```
Comment (jogold) [inline packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]:
```suggestion
   * The ACM certificate; Has to be in us-east-1
```

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 4 above)
Comment (jogold) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: I presume that this is to remove the `http://` from the website url attribute

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Nice. good to know. thanks

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 4 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Wouldn't that be `redirectBucket.bucketDomainName`?

Comment (jogold) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: `bucketDomainName` points to the REST API endpoint, not the website endpoint. The REST API endpoint doesn't support redirects.

`mystack-mybucket-kdwwxmddtr2g.s3.amazonaws.com` vs `http://mystack-mybucket-kdwwxmddtr2g.s3-website-us-east-2.amazonaws.com/`

https://docs.aws.amazon.com/AmazonS3/latest/dev/WebsiteEndpoints.html#WebsiteRestEndpointDiff

@hoegertn may be add a comment explaining this line.

Comment (hoegertn) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: the bucketDomainName is the name for S3 operations but not the website domain

File: packages/@aws-cdk/aws-route53-patterns/README.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+---
+<!--END STABILITY BANNER-->
+
+This library contains commonly used patterns for Route53.
+
+
+## HTTPS Redirect
+
+This construct allows creating a simple HTTP->HTTPS and domainA->domainB redirect using CloudFront and S3.
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-route53-patterns/README.md]: Add an snippet on how to create an HTTP=>HTTPS redirect. It's not clear from this.

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 4 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: Should we simply add another attribute to `s3.Bucket`? Sounds like something others can benefit from...

File: packages/@aws-cdk/aws-route53-patterns/README.md (hunk 10 above)
Comment (hoegertn) [reply packages/@aws-cdk/aws-route53-patterns/README.md]: I will reword this sentence. It is doing a domain redirect and http->https redirect for domainA. You cannot use this construct to redirect http->https on your app domain, as you can only have of DNS entry for your domain.

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 5 above)
Comment (hoegertn) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: As it is only doing redirects this should be a non-issue. But if the need arises, this is a  quick change.

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: @hoegertn what do you think?

File: packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts (hunk 4 above)
Comment (eladb) [reply packages/@aws-cdk/aws-route53-patterns/lib/website-redirect.ts]: @hoegertn care to raise an issue (and make sure to reference this usage so we can track it down).

File: packages/@aws-cdk/aws-route53-patterns/README.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+```ts
+new HttpsRedirect(stack, 'Redirect', {
+  domainName: 'foo.example.com',
```
Comment (jogold) [inline packages/@aws-cdk/aws-route53-patterns/README.md]: This is now `recordNames`

**Comment by jogold** [comment]
> @eladb Something strange is happening on build. If I remove the dependency in `decdk` it fails and if I add it it does not find it. What do I have to do?

Seems like your `package-lock.json` is not in sync anymore with your `package.json`:
- update package version to `1.8.0` in `package.json`
- update all cdk dependencies in your `package.json` to `^1.8.0` (remove unused `@aws-cdk/region-info`)
- inside `aws-route53-patterns`: `rm -rf package-lock.json && npm i`

**Pull request merged**